xsdk 0.5.0

balay opened the pull request:
~~This PR is for early review/comments, and is not yet ready for merge~~

~~Its has commits from other PRs [and workarounds] that won't be in the final set of changes.~~

~~https://github.com/spack/spack/pull/13661~~
~~[trilinos changes]~~
~~[metis changes]~~

~~i.e the PR is likely to only have changes for the following - and I would appreciate feedback/suggestions on these changes.~~
 
~~var/spack/repos/builtin/packages/alquimia/package.py~~
~~var/spack/repos/builtin/packages/pflotran/package.py~~
~~var/spack/repos/builtin/packages/xsdk/package.py~~

we are hoping to get this done by nov 15th

cc: @ulrikeyang
cc: @jwillenbring

adamjstewart commented on var/spack/repos/builtin/packages/xsdk/package.py:
```diff
@@ -18,91 +19,128 @@ class Xsdk(BundlePackage):
 
-    depends_on('tasmanian@develop+xsdkflags+blas~openmp', when='@develop')
-    depends_on('tasmanian@develop+xsdkflags+blas+cuda+magma~openmp', when='@develop +cuda')
+    tasmanian_openmp = '~openmp' if sys.platform == 'darwin' else '+openmp'
```
Couple of comments here. First off, it would be more correct to do something like `self.spec.satisfies('%clang platform=darwin')` since other compilers have OpenMP support, but I don't think there's any way to do that here. Second of all, you _can_ actually get OpenMP working with Apple Clang if you add a dependency on `llvm-openmp` and use the right build flags. See `py-scikit-learn` and #13669 for examples of this.

balay replied:
Thanks! for one - we can't access self.spec in the class constructor - hence this change.
And `lvm-openmp` option looks good. So I think we should use this.

But I get build errors with tasmanian cc: @mkstoyanov 

```
==> [2019-11-11-13:26:46.695709] 'cmake' '/var/folders/xk/v0fd809s2y7b3x8bp05l2nww0000gq/T/balay/spack-stage/spack-stage-tasmanian-7.0-u55sduzuuzqw77kbq42hjchs72qx7cpp/spack-src' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:PATH=/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/tasmanian-7.0-u55sduzuuzqw77kbq42hjchs72qx7cpp' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_FIND_FRAMEWORK:STRING=LAST' '-DCMAKE_FIND_APPBUNDLE:STRING=LAST' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=FALSE' '-DCMAKE_INSTALL_RPATH:STRING=/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/tasmanian-7.0-u55sduzuuzqw77kbq42hjchs72qx7cpp/lib;/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/tasmanian-7.0-u55sduzuuzqw77kbq42hjchs72qx7cpp/lib64;/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/llvm-openmp-9.0.0-kkckpcgbwofd7nmpry7nq2bdgq3gvdec/lib' '-DCMAKE_PREFIX_PATH:STRING=/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/mpich-3.3.1-dsalg5laudrzwm2svunxhcvq5xsrnigh;/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/openblas-0.3.7-5laehbhhughfsh5rv7iosk56j3lcecof;/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/llvm-openmp-9.0.0-kkckpcgbwofd7nmpry7nq2bdgq3gvdec' '-DUSE_XSDK_DEFAULTS:BOOL=ON' '-DXSDK_ENABLE_PYTHON:BOOL=OFF' '-DTPL_ENABLE_MPI:BOOL=ON' '-DXSDK_ENABLE_OPENMP:BOOL=ON' '-DTPL_ENABLE_BLAS:BOOL=ON' '-DXSDK_ENABLE_CUDA:BOOL=OFF' '-DTPL_ENABLE_MAGMA:BOOL=OFF' '-DXSDK_ENABLE_FORTRAN:BOOL=OFF'
-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- Check for working CXX compiler: /Users/balay/spack/lib/spack/env/clang/clang++
-- Check for working CXX compiler: /Users/balay/spack/lib/spack/env/clang/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindOpenMP.cmake:477 (find_package_handle_standard_args)
  Config/CMakeIncludes/sanity_check_and_xsdk.cmake:75 (find_package)
  CMakeLists.txt:111 (include)
```

mkstoyanov replied:
The error is simply that OpenMP option has been turned `ON`, but CMake cannot find OpenMP for this particular compiler.

Apparently the `lvm-openmp` dependencie was not registered.

adamjstewart replied:
You may also need to add a few additional flags, like `-Xpreprocessor -fopenmp` and `-lomp`. I also needed to add `-I/path/to/llvm-openmp/include` and `-L/path/to/llvm-openmp/lib` for other packages.

mkstoyanov replied:
@adamjstewart CMake is supposed to take care of the OpenMP flags and libs, if CMake doesn't find those, then it would be very hard to pass the flags manually.

adamjstewart replied:
@mkstoyanov CMake's FindOpenMP.cmake does not seem sufficient to find llvm-openmp when using Apple Clang. See https://github.com/intel/mkl-dnn/issues/591 for more details, and see #13669 for an example of a CMake package where I had to manually override this.

mkstoyanov replied:
@adamjstewart this is good to know for the future, but I think the change is too much for a release that is overdue. It will take quite some time to retest this if I make the changes now.

adamjstewart commented on var/spack/repos/builtin/packages/xsdk/package.py:
```diff
@@ -18,91 +19,128 @@ class Xsdk(BundlePackage):
-    depends_on('plasma@18.11.1:', when='@develop %gcc@6.0:')
+    depends_on('plasma@19.8.1:', when='@develop %gcc@6.0:')
+    depends_on('plasma@19.8.1:', when='@0.5.0 %gcc@6.0:')
     depends_on('plasma@18.11.1:', when='@0.4.0 %gcc@6.0:')
```
What exactly does this mean? Plasma isn't needed when building with older versions of GCC?

balay replied:
Here is the case were we need to enable plasma only when we know the compiler supports openmp-4.5+. So for now - we are enabling plasma  for `gcc@6.0:` which is likely a subset of compilers that can satisfy the above requirement

cc: @luszczek

adamjstewart commented on var/spack/repos/builtin/packages/xsdk/package.py:
```diff
@@ -18,91 +19,128 @@ class Xsdk(BundlePackage):
+    depends_on('amrex@19.08', when='@0.5.0 %intel')
+    depends_on('amrex@19.08', when='@0.5.0 %gcc')
     depends_on('amrex@18.10.1', when='@0.4.0 %intel')
     depends_on('amrex@18.10.1', when='@0.4.0 %gcc')
```
What exactly does this mean? Amrex isn't needed when building with other compilers?

balay replied:
Here I'd like to exclude clang - but couldn't figure out spack syntax for it. [there is a conflict on clang in amrex].

adamjstewart replied:
Do you want to prevent people from building `xsdk` with clang, or just `amrex`?

balay replied:
Most packages build fine with clang. I  want to exclude only `amrex` as it won't build with clang.

adamjstewart commented on var/spack/repos/builtin/packages/xsdk/package.py:
```diff
@@ -111,8 +149,26 @@ class Xsdk(BundlePackage):
+    depends_on('ginkgo@develop ~openmp', when='@develop +ginkgo')
+    depends_on('ginkgo@develop ~openmp+cuda', when='@develop +ginkgo +cuda')
+    depends_on('ginkgo@1.1.0 ~openmp', when='@0.5.0 +ginkgo')
+    depends_on('ginkgo@1.1.0 ~openmp+cuda', when='@0.5.0 +cuda +ginkgo')
```
Why not use the same OpenMP treatment above?

balay replied:
At some point we have add openmp variant to xsdk and consistently enable/disable it for all packages.

Then there is this issue of different openmp versions that we need to figure out [some packages need openmp version 4.5+]. So we haven't worked out this yet. Does spack has some infrastructure for detecting and using this info in pacakges?

```
diff --git a/var/spack/repos/builtin/packages/xsdk/package.py b/var/spack/repos/builtin/packages/xsdk/package.py
index 354db12d6..c13f33100 100644
--- a/var/spack/repos/builtin/packages/xsdk/package.py
+++ b/var/spack/repos/builtin/packages/xsdk/package.py
@@ -156,10 +153,9 @@ class Xsdk(BundlePackage):
     depends_on('phist@1.8.0 ~fortran ~scamac ~openmp ~host', when='@0.5.0 +phist')
     depends_on('phist@1.7.5 ~fortran ~scamac ~openmp ~host', when='@0.4.0 +phist')
 
-    depends_on('ginkgo@develop ~openmp', when='@develop +ginkgo')
-    depends_on('ginkgo@develop ~openmp+cuda', when='@develop +ginkgo +cuda')
-    depends_on('ginkgo@1.1.0 ~openmp', when='@0.5.0 +ginkgo')
-    depends_on('ginkgo@1.1.0 ~openmp+cuda', when='@0.5.0 +cuda +ginkgo')
+    depends_on('ginkgo +cuda', when='+ginkgo +cuda')
+    depends_on('ginkgo@develop', when='@develop +ginkgo')
+    depends_on('ginkgo@1.1.0', when='@0.5.0 +ginkgo')
 
     depends_on('py-libensemble@develop+petsc4py', when='@develop +libensemble')
     depends_on('py-libensemble@0.5.2+petsc4py', when='@0.5.0 +libensemble')

```

```
diff --git a/var/spack/repos/builtin/packages/ginkgo/package.py b/var/spack/repos/builtin/packages/ginkgo/package.py
index a0de875a5..06affc022 100644
--- a/var/spack/repos/builtin/packages/ginkgo/package.py
+++ b/var/spack/repos/builtin/packages/ginkgo/package.py
@@ -23,7 +23,7 @@ class Ginkgo(CMakePackage, CudaPackage):
 
     variant('shared', default=True, description='Build shared libraries')
     variant('full_optimizations', default=False, description='Compile with all optimizations')
-    variant('openmp', default=sys.platform != 'darwin',  description='Build with OpenMP')
+    variant('openmp', default=True,  description='Build with OpenMP')
     variant('develtools', default=False, description='Compile with develtools enabled')
     variant('build_type', default='Release',
             description='The build type to build',
@@ -31,6 +31,7 @@ class Ginkgo(CMakePackage, CudaPackage):
 
     depends_on('cmake@3.9:', type='build')
     depends_on('cuda@9:',    when='+cuda')
+    depends_on('llvm-openmp', when='+openmp %clang platform=darwin')
 
     conflicts('%gcc@:5.2.9')

```
I could change the above to enable openmp [similar to tasmanian] -  would have to redo all testing

I get same failure on OSX+clang. Perhaps I'm doing something wrong here..


```
==> Executing phase: 'cmake'
==> [2019-11-11-14:06:57.630744] 'cmake' '/var/folders/xk/v0fd809s2y7b3x8bp05l2nww0000gq/T/balay/spack-stage/spack-stage-ginkgo-1.1.0-b6ggjffdhcohmns4v2soomf65ipykpjd/spack-src' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:PATH=/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/ginkgo-1.1.0-b6ggjffdhcohmns4v2soomf65ipykpjd' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_FIND_FRAMEWORK:STRING=LAST' '-DCMAKE_FIND_APPBUNDLE:STRING=LAST' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=FALSE' '-DCMAKE_INSTALL_RPATH:STRING=/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/ginkgo-1.1.0-b6ggjffdhcohmns4v2soomf65ipykpjd/lib;/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/ginkgo-1.1.0-b6ggjffdhcohmns4v2soomf65ipykpjd/lib64;/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/llvm-openmp-9.0.0-kkckpcgbwofd7nmpry7nq2bdgq3gvdec/lib' '-DCMAKE_PREFIX_PATH:STRING=/Users/balay/spack/opt/spack/darwin-catalina-x86_64/clang-11.0.0-apple/llvm-openmp-9.0.0-kkckpcgbwofd7nmpry7nq2bdgq3gvdec' '-DGINKGO_BUILD_CUDA=OFF' '-DGINKGO_BUILD_OMP=ON' '-DBUILD_SHARED_LIBS=ON' '-DGINKGO_JACOBI_FULL_OPTIMIZATIONS=OFF' '-DGINKGO_DEVEL_TOOLS=OFF' '-DGINKGO_BUILD_BENCHMARKS=OFF' '-DGINKGO_BUILD_DOC=OFF' '-DGINKGO_BUILD_EXAMPLES=OFF' '-DGINKGO_BUILD_TESTS=OFF'
-- The C compiler identification is AppleClang 11.0.0.11000033
-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- Check for working C compiler: /Users/balay/spack/lib/spack/env/clang/clang
-- Check for working C compiler: /Users/balay/spack/lib/spack/env/clang/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Users/balay/spack/lib/spack/env/clang/clang++
-- Check for working CXX compiler: /Users/balay/spack/lib/spack/env/clang/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test Ginkgo_C_COVERAGE_SUPPORTED
-- Performing Test Ginkgo_C_COVERAGE_SUPPORTED - Success
-- Performing Test Ginkgo_C_TSAN_SUPPORTED
-- Performing Test Ginkgo_C_TSAN_SUPPORTED - Failed
-- Performing Test Ginkgo_C_ASAN_SUPPORTED
-- Performing Test Ginkgo_C_ASAN_SUPPORTED - Success
-- Performing Test Ginkgo_CXX_COVERAGE_SUPPORTED
-- Performing Test Ginkgo_CXX_COVERAGE_SUPPORTED - Success
-- Performing Test Ginkgo_CXX_TSAN_SUPPORTED
-- Performing Test Ginkgo_CXX_TSAN_SUPPORTED - Failed
-- Performing Test Ginkgo_CXX_ASAN_SUPPORTED
-- Performing Test Ginkgo_CXX_ASAN_SUPPORTED - Success
-- Looking for C++ include cxxabi.h
-- Looking for C++ include cxxabi.h - found
-- Could NOT find PAPI (missing: PAPI_LIBRARY PAPI_INCLUDE_DIR) 
CMake Error at /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindOpenMP.cmake:477 (find_package_handle_standard_args)
  omp/CMakeLists.txt:1 (find_package)


-- Configuring incomplete, errors occurred!
```

cc: @hartwiganzt

adamjstewart replied:
Currently Spack doesn't have any info on which versions of OpenMP a particular compiler supports. We have a hacky way of detecting C++?? support, but not OpenMP version. Ideally there would be a way to require a compiler with both.

balay commented:
All the supporting PRs are already merged. I've rebased this PR over latest develop - and collapsed the commits [and updated the commit message]

I don't anticipate any more changes here. So I've removed the WIP tag  [However keeping the  'don't-yet-merge' label for now.]

Wrt current feedback - I think most of them are things we need to improve in the future.

@adamjstewart if there are any show stoppers here - let us know. Thanks!

adamjstewart commented on var/spack/repos/builtin/packages/xsdk/package.py:
```diff
@@ -111,8 +150,27 @@ class Xsdk(BundlePackage):
+    depends_on('ginkgo@1.1.0 ~openmp+cuda', when='@0.5.0 +cuda +ginkgo')
+
+    depends_on('py-libensemble@develop+petsc4py', when='@develop +libensemble')
+    depends_on('py-libensemble@0.5.2+petsc4py', when='@0.5.0 +libensemble')
```
These might be `type=('build', 'run')` dependencies.

balay replied:
Hm - `py-libensemble` is another numerical package and not really a build dependency

adamjstewart replied:
Now that I think about it, it really doesn't matter what dependency type most of these packages are, as there is no code in `xsdk` to link to. In that case, this can just be `type='run'` since it needs to be present in `PYTHONPATH` at run-time.

balay replied:
ok - pushed

acfisher commented:
I got this error when trying to spack install xsdk@0.5.0:
xsdk requires petsc variant +trilinos, but spec asked for ~trilinos

Are we planning to disable trilinos in this version?

balay commented:
> I got this error when trying to spack install xsdk@0.5.0:
> xsdk requires petsc variant +trilinos, but spec asked for ~trilinos
> 
> Are we planning to disable trilinos in this version?

Hm - I'm not sure why this comes up.

` spack spec xsdk` works but `spack spec xsdk@0.5.0` does not. The default is `+trilinos` so why does this change to `~trilinos` when we change from `xsdk` to `xsdk@0.5.0`?

Perhaps this is some concretizer issue?

cc: @adamjstewart

balay commented:
I'm having to do the following for the above 2 options to be equivalent. 
```
[balay@p1 ~]$ spack spec xsdk > a.log
[balay@p1 ~]$ spack spec xsdk@0.5.0%gcc@9.2.1 +trilinos+dealii > b.log
[balay@p1 ~]$ diff a.log b.log
3c3
< xsdk
---
> xsdk@0.5.0%gcc@9.2.1+dealii+trilinos
[balay@p1 ~]$ 
```

acfisher commented:
I'm also failing to get petsc to build on lassen with xl.  It looks like it it aborting due to warnings.
 >> 25    ***** WARNING: F77 (set to /usr/WS1/fisher47/spack/lib/spack/env/xl/xlf) found in environment variables - ignoring
>> 57    ***** WARNING: Using default optimization CUDA flags -O
We may have to give up on xl if we keep uncovering problems like this.

balay commented:
> I'm also failing to get petsc to build on lassen with xl. It looks like it it aborting due to warnings.

Perhaps we can move these issues to https://github.com/xsdk-project/xsdk-issues/issues
[if you can provide logs there - I can take a look]

balay commented:
I've collapsed the commits [updated the commit message] - and rebased over latest develop. And also removed the don't-merge-yet tag.

Will run a few test now.

acfisher commented:
I'm having other problems with the concretizer on Lassen as well.  For some reason when I try to build petsc with +batch under xsdk it won't override the default:
xsdk requires petsc variant +batch, but spec asked for ~batch

As far as I know noting in xsdk requires petsc to build with ~batch, but it sure is acting like it is required. @adamjstewart

balay commented:
@acfisher  `./bin/spack spec xsdk ^petsc+batch` works fine on my linux box - so not sure what command you are using on Lassen.

either way - perhaps we should create a new issue for it - and merge this  pull request in.
